### ISSUE TITLE
Add ceiling division to math library

### ISSUE THREAD

**Issue by nlordell** [OpenZeppelin/openzeppelin-contracts#2680]
**🧐 Motivation**

This FR proposes adding a `ceilDiv` method to the `Math.sol` library for performing division that rounds up instead of truncating (standard integer division).

**📝 Details**

Ceiling division is fairly easy to implement, and can either be done with:
- `(a + b - 1) / b`
- `a / b + uint(a % b != 0)`

The second implementation is probably desired, as it does not overflow for large `a` and `b` where the result would fit in a 256-bit integer.

**Comment by frangio** [comment]
@nlordell This sounds reasonable but would you mind sharing an example of a situation where you would use this method?

**Comment by nlordell** [comment]
Sure! The particular one we used this for (which prompted upstreaming this change) was for computing executed buy amounts given an input sell amount and rational price.

For example, if you were to compute a buy amount that a user would receive given some input sell amount and price fraction:
```sol
function computeBuyAmount(
    uint256 sellAmount,
    uint256 priceNumerator,
    uint256 priceDenominator
) returns (uint256) {
  return (sellAmount * priceNumerator).ceilDiv(priceDenominator);
}
```

Implementing this rounding up instead of down makes rounding go in favour of the user. For example, if a user would sell a very small amount with a price <1 (i.e. `priceNumerator < priceDenominator`), then the buy amount would compute to 0.

**Comment by frangio** [comment]
Hm, I think there is probably a place for `ceilDiv` but I'm not sure that's a good one.

It's generally recommended to always round in favor of the contract, for security reasons. Otherwise, it can leave the contract with a "negative" balance, which can result in future transactions failing due to insufficient balance.

**Comment by nlordell** [comment]
> It's generally recommended to always round in favor of the contract, for security reasons.

So specifically, in our case, we are dealing with orders that are signed by users off-chain and executed by a third party. This means that, rounding in favour of the user makes more sense here as they are "passively" trading, and could get taken advantage of by the third party if rounding was done in favour of the SC. The following could happen to partially fillable orders using truncated division in that case (in pseudo-code):

```solidity
uint256 epsilonSellAmount = // largest sell amount that would compute the a buy amount of 0 because of truncating division
while (orderNotCompletelyFilled()) {
    partiallyFillOrder(epsilonSellAmout);
}
```

This isn't super convenient, as this would be very gas expensive, but this essentially would drain the entire order sell amount of a user and provide 0 buy amount. Additionally, the epsilon sell amount could be worth ~1 cent if tokens like GUSD (with `decimals = 2`) are involved.

> Otherwise, it can leave the contract with a "negative" balance, which can result in future transactions failing due to insufficient balance.

In our case, the SC doesn't rely on existing balances so this isn't a concern. For an AMM that pools balances, I completely agree with this statement.

However, even in that case, there is still a place for ceiling division. Imagine an AMM where you can specify an order with a maximum sell amount and limit price fraction. Ceiling division could be used to compute the minimum buy amount given some executed sell amount:

```solidity
function swap(
    uint256 maxSellAmount,
    uint256 priceNumerator,
    uint256 priceDenominator
) returns (uint256) {
  (uint256 sellAmount, uint256 buyAmount) = swapUpTo(maxSellAmount);
  // here sellAmount <= maxSellAmount
  uint256 minBuyAmount = (sellAmount * priceNumerator).ceilDiv(priceDenominator);
  require(buyAmount >= minBuyAmount);
}
```

Here the ceiling division is used to protect the user from price movement, while making sure that the "minimum buy amount" computation still rounds in their favour (regardless of how the actual buy amount is computed).

**Comment by nlordell** [comment]
As a side note:

We love and use OZ contracts, and added this to the math library in our project. The issue and PR was more in the spirit: "hey! we added this to our vendorred OZ math library! feel free to add it if you think it's useful and will help others". That being said, if you think its scope is too narrow to be included here, that's a totally valid reason to close this issue (and PR), for which I don't have any strong arguments against.

Just wanted to share some code we found useful :smile:.
